llama_index/evaluation/multi_modal/faithfulness.py

```diff
@@ -0,0 +1,222 @@
+    "is supported by the visual as well as textual context information.\n"
+    "You need to answer with either YES or NO.\n"
+    "Answer YES if any of the image(s) and textual context supports the information, even "
+    "if most of the context is unrelated. "
+    "Some examples are provided below with only text context, but please do use\n"
+    "any images for context if they are provided.\n\n"
+    "Information: Apple pie is generally double-crusted.\n"
+    "Context: An apple pie is a fruit pie in which the principal filling "
+    "ingredient is apples. \n"
+    "Apple pie is often served with whipped cream, ice cream "
+    "('apple pie à la mode'), custard or cheddar cheese.\n"
+    "It is generally double-crusted, with pastry both above "
+    "and below the filling; the upper crust may be solid or "
+    "latticed (woven of crosswise strips).\n"
+    "Answer: YES\n"
+    "Information: Apple pies tastes bad.\n"
+    "Context: An apple pie is a fruit pie in which the principal filling "
+    "ingredient is apples. \n"
+    "Apple pie is often served with whipped cream, ice cream "
+    "('apple pie à la mode'), custard or cheddar cheese.\n"
+    "It is generally double-crusted, with pastry both above "
+    "and below the filling; the upper crust may be solid or "
+    "latticed (woven of crosswise strips).\n"
+    "Answer: NO\n"
+    "Information: {query_str}\n"
+    "Context: {context_str}\n"
+    "Answer: "
+)
+
+DEFAULT_REFINE_TEMPLATE = PromptTemplate(
+    "We want to understand if the following information is present "
+    "in the context information: {query_str}\n"
+    "We have provided an existing YES/NO answer: {existing_answer}\n"
+    "We have the opportunity to refine the existing answer "
+    "(only if needed) with some more context below.\n"
+    "------------\n"
+    "{context_msg}\n"
+    "------------\n"
+    "If the existing answer was already YES, still answer YES. "
+    "If the information is present in the new context, answer YES. "
+    "Otherwise answer NO.\n"
+)
+
+
+class MultiModalFaithfulnessEvaluator(BaseEvaluator):
+    """Multi-Modal Faithfulness evaluator.
+
+    Evaluates whether a response is faithful to the contexts
+    (i.e. whether the response is supported by the contexts or hallucinated.)
+
+    This evaluator only considers the response string and the list of context strings.
+
+    Args:
+        multi_modal_llm(Optional[MultiModalLLM]):
+            The Multi-Modal LLM Judge to use for evaluations.
+        raise_error(bool): Whether to raise an error when the response is invalid.
+            Defaults to False.
+        eval_template(Optional[Union[str, BasePromptTemplate]]):
+            The template to use for evaluation.
+        refine_template(Optional[Union[str, BasePromptTemplate]]):
+            The template to use for refining the evaluation.
+    """
+
+    def __init__(
+        self,
+        multi_modal_llm: MultiModalLLM | None = None,
```

Comment:
Pretty sure python3.8 doesn't support the `|` type operator -- better to use `Union` for now

Reply:
For some reason, my linter suggested this, which surprised me given that I saw we use `Union` or `Optional`. Okay i'll use `Union` and see if it doesn't complain (i'm on 3.10)